networking/notes/shelley/CommunicationPatterns.md

```diff
@@ -0,0 +1,283 @@
+need to request that data from peers, so a hybrid request/response and pub/sub
+model is needed. Transaction, delegation, and update submission also require
+the active request/response approach, because edge nodes behind a firewall
+and/or NAT must be able to do this, but such a node cannot have any
+subscribers. Even publishing a new block (slot leader) must work by
+request/response, for edge nodes must be capable of doing this as well.
+
+### Porting the Inv/Req/Data relay system
+
+We believe the key/value distinction is still important: you can send a key
+identifying the data and the receiver can request or reject the payload. This
+will probably save a lot of wasted traffic as a piece of data can reach a node
+by multiple paths.
+
+When a piece of data is being published (like a new block or a transaction,
+see request/response) the key/value can be skipped and the whole payload can
+be sent at once, because we know the receiver won't have it. The key/value
+distinction is only useful when relaying, and so is relevant only to pub/sub.
+
+When a datum is to be published (that's to say, republished, or relayed), its
+key will be sent to all subscribers, and the subscribers can asynchronously
+request the data for it or indicate that it will not request it. This could
+permit a TCP-style flow control algorithm: if the publisher notices too much
+"un-ACKed" data (keys for which the particular subscriber has not returned a
+yes or no answer) then it can attenuate the flow.
+
+To avoid cycles in relaying, a node must remember keys which have been
+published to it for a certain amout of time. If such a key comes up again,
+it will send a negative ACK meaning "I don't want the payload" and it will not
+be relayed. A datum is relayed only if the payload was requested.
+The current implementation uses the blockchain and transaction mempool to
+determine whether something should be requested and relayed; the diffusion
+layer will use its own cache.
+
+Here's a crude state machine for the relay cache.
+
+States:
+
+  1. `k` and `v` are unknown
+  2. `k` is known but `v` is unknown
+  3. `k` and `v` are known
+
+Transitions:
+
+  - `k` is published to us
+    If in state 1, ask the publisher for `v`, enter state 2.
+    If in state 2, reject the offer from the publisher, remain in state 2.
+    If in state 3, reject the offer from the publisher, remain in state 3.
+  - `v` is published to us
+    If in state 1, remain in state 1. This should not happen.
+    If in state 2, remember `v` and relay `k` and `v` to subscribers, enter
+      state 3.
+    If in state 3, remain in state 3. This should not happen.
+  - TTL on the key expires.
+    If in state 1, this makes no sense, since we don't have the key.
+    If in state 2, forget `v` and enter state 1.
+      This is dodgy. It means the publisher failed to deliver before TTL.
+    If in state 3, forget `k` and `v` and enter state 1.
```

Comment:
@coot said

> We could also relay `k` not yet knowing `v` and then queue the requests for `v` until we know it. But this would complicate matters when we fail to get `v`, so it's better to start with the proposed machine.
> 
> There should be a fallback when the request for `v` fails, anyway.  So maybe we should have three states for `v`: `unknown`, `failed` and `known`?